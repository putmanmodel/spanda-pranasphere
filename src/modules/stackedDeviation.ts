export type ToneSnapshot = {
  pole: number;            // From tonePairs
  code: string;            // Â¡XX.YYY[z] format
  deviation: number;       // From detectDeviation
  timestamp: number;
};

export type ArcInsight = {
  arcType: "flat" | "swing" | "surge" | "collapse";
  triggerLevel: number;
  peakPole?: number;
  deviationRange: number;
  avgDeviation: number;
  trend: "rising" | "falling" | "stable";
};

/**
 * Analyzes a sequence of tone snapshots and outputs an ArcInsight summary.
 */
export function analyzeStackedDeviation(history: ToneSnapshot[]): ArcInsight {
  if (history.length === 0) {
    return {
      arcType: "flat",
      triggerLevel: 0,
      deviationRange: 0,
      avgDeviation: 0,
      trend: "stable"
    };
  }

  const deviations = history.map(h => h.deviation);
  const peak = Math.max(...deviations);
  const trough = Math.min(...deviations);
  const avg = deviations.reduce((a, b) => a + b, 0) / deviations.length;
  const range = peak - trough;

  const first = deviations[0];
  const last = deviations[deviations.length - 1];
  const trend =
    last > first + 0.2 ? "rising" :
    last < first - 0.2 ? "falling" :
    "stable";

  const arcType =
    range > 1.5 ? "surge" :
    range > 0.7 ? "swing" :
    avg < -0.7 ? "collapse" :
    "flat";

  return {
    arcType,
    triggerLevel: Math.abs(peak) > Math.abs(trough) ? peak : trough,
    peakPole: history.find(h => h.deviation === (Math.abs(peak) > Math.abs(trough) ? peak : trough))?.pole,
    deviationRange: range,
    avgDeviation: avg,
    trend
  };
}