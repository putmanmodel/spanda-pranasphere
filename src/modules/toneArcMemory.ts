type PoleStats = {
  total: number;
  count: number;
  decay: number;
};

type MoodField = {
  poleAverages: Record<number, PoleStats>;
  lastUpdate: number;
  trend: "rising" | "falling" | "stable";
};

const defaultDecay = 0.95;
const poleMemory: MoodField = {
  poleAverages: {},
  lastUpdate: Date.now(),
  trend: "stable",
};

export function updateToneField(pole: number, deviation: number, decayRate = defaultDecay) {
  const stats = poleMemory.poleAverages[pole] || { total: 0, count: 0, decay: decayRate };
  stats.total = stats.total * stats.decay + deviation;
  stats.count = stats.count + 1;
  stats.decay = decayRate;
  poleMemory.poleAverages[pole] = stats;

  const previousAvg = getOverallAverage();
  const newAvg = getOverallAverage();

  const change = newAvg - previousAvg;
  poleMemory.trend =
    change > 0.1 ? "rising" :
    change < -0.1 ? "falling" :
    "stable";

  poleMemory.lastUpdate = Date.now();
}

export function getOverallAverage(): number {
  const values = Object.values(poleMemory.poleAverages);
  if (values.length === 0) return 0;

  const weightedSum = values.reduce((acc, v) => acc + v.total, 0);
  const count = values.reduce((acc, v) => acc + v.count, 0);
  return weightedSum / count;
}

export function getMoodField(): MoodField {
  return {
    poleAverages: { ...poleMemory.poleAverages },
    lastUpdate: poleMemory.lastUpdate,
    trend: poleMemory.trend,
  };
}

export function resetMoodField() {
  for (const key in poleMemory.poleAverages) delete poleMemory.poleAverages[key];
  poleMemory.lastUpdate = Date.now();
  poleMemory.trend = "stable";
}