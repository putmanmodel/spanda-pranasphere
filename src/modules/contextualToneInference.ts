import { ToneSnapshot } from "./stackedDeviation";
import { MoodFieldSnapshot } from "../types";

export type ContextualToneResult = {
  resonance: "aligned" | "neutral" | "clashing";
  modulation: "amplify" | "dampen" | "redirect" | "none";
  notes?: string;
};

/**
 * Infers the contextual alignment of an incoming tone based on recent mood field trends.
 */
export function inferContextualTone(
  snapshot: ToneSnapshot,
  mood: MoodFieldSnapshot
): ContextualToneResult {
  const poleData = mood.poleAverages[snapshot.pole];

  if (!poleData) {
    return {
      resonance: "neutral",
      modulation: "none",
      notes: "No prior data for this pole. Neutral stance assumed."
    };
  }

  const avg = poleData.total / poleData.count;
  const diff = snapshot.deviation - avg;

  if (Math.abs(diff) < 0.2) {
    return {
      resonance: "aligned",
      modulation: "amplify",
      notes: "Tone closely matches pole's recent mood average."
    };
  }

  if (Math.abs(diff) > 1.0 && Math.sign(diff) !== Math.sign(avg)) {
    return {
      resonance: "clashing",
      modulation: "redirect",
      notes: "Tone sharply contradicts recent pole average — possible disruption."
    };
  }

  return {
    resonance: "neutral",
    modulation: "dampen",
    notes: "Tone diverges mildly from trend — recommend de-emphasizing."
  };
}