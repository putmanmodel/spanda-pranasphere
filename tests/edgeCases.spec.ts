import { analyzeStackedDeviation } from "../src/modules/stackedDeviation";
import cases from "../fixtures/edge-cases.json";

describe("Edge-Case Arc Analysis", () => {
  for (const c of cases) {
    test(c.id, () => {
      const arc = analyzeStackedDeviation(c.snapshots);
      expect(arc).toMatchSnapshot();
    });
  }
});