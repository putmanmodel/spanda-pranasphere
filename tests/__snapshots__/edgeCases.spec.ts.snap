// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Edge-Case Arc Analysis clashing-signals 1`] = `
{
  "arcType": "swing",
  "avgDeviation": -0.025000000000000022,
  "deviationRange": 1.45,
  "peakPole": 4,
  "trend": "falling",
  "triggerLevel": -0.75,
}
`;

exports[`Edge-Case Arc Analysis flatline-no-deviation 1`] = `
{
  "arcType": "flat",
  "avgDeviation": 0,
  "deviationRange": 0,
  "peakPole": 2,
  "trend": "stable",
  "triggerLevel": 0,
}
`;

exports[`Edge-Case Arc Analysis overamplification-drift 1`] = `
{
  "arcType": "swing",
  "avgDeviation": 1.2333333333333334,
  "deviationRange": 0.7000000000000001,
  "peakPole": 5,
  "trend": "rising",
  "triggerLevel": 1.6,
}
`;

exports[`Edge-Case Arc Analysis sarcasm-flip 1`] = `
{
  "arcType": "surge",
  "avgDeviation": -0.19999999999999996,
  "deviationRange": 2,
  "peakPole": 1,
  "trend": "falling",
  "triggerLevel": -1.2,
}
`;
